assets/js/modules/analytics-4/utils/report-normalization.js

```diff
@@ -0,0 +1,59 @@
+/**
+ * Report normalization utilities.
+ *
+ * Site Kit by Google, Copyright 2022 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/**
+ * External dependencies
+ */
+import castArray from 'lodash/castArray';
+import isPlainObject from 'lodash/isPlainObject';
+import memize from 'memize';
+
+/**
+ * Normalizes report options.
+ *
+ * @since 1.17.0
+ *
+ * @param {Object} options Report options object.
+ * @return {Object} Normalized options object.
+ */
+export const normalizeReportOptions = memize(
+	( { metrics, dimensions, ...options } = {} ) => {
+		// TODO: build this out to normalize all options used.
+		return {
+			metrics: normalizeMetrics( metrics ),
+			dimensions: normalizeDimensions( dimensions ),
+			...options,
+		};
+	}
+);
+
+const normalizeMetrics = ( metrics ) => {
+	return castArray( metrics )
+		.map( ( metric ) =>
+			typeof metric === 'string' ? { expression: metric } : metric
+		)
+		.filter( ( metric ) => isPlainObject( metric ) );
+};
+
+const normalizeDimensions = ( dimensions ) => {
+	return castArray( dimensions )
+		.map( ( dimension ) =>
+			typeof dimension === 'string' ? { name: dimension } : dimension
+		)
+		.filter( ( dimension ) => isPlainObject( dimension ) );
```

Comment:
We probably need to [review the schema for the various options](https://developers.google.com/analytics/devguides/reporting/data/v1/rest/v1beta/properties/runReport) to make sure these work as expected. There's probably also some collaboration to do between this issue and #6172 as that's where the actual report options will be passed through. We generally aim to keep things consistent between our API and the service parameters. Our endpoint has historically been a bit more flexible though. Let's look into this and discuss :)

Reply:
Should we maybe set #6172 as a blocker for this issue?

Reply:
Yes I think so. Originally I was thinking these could happen independently but this part does need solidifying to inform how it should be used by the client 👍 I'll do that now. It might also be good for you to take that issue as well since you have important relevant context here. I'll assign you for now but if you're not comfortable taking it just let me know.

Reply:
@aaemnnosttv I had a chat with @techanvil to update the normalize functions as well as the validation functions to match the GA4 requirements.